### PR TITLE
[Clang][AArch64] Expose compatible SVE intrinsics with only +sme

### PR DESCRIPTION
**Pull request by sdesmalen-arm**
This allows code with SVE intrinsics to be compiled with +sme,+nosve,                                                                                                                                                                                                                                                                                                   
assuming the encompassing function is in the correct mode (see #93802)

### CODE REVIEW

**Review by paulwalker-arm — COMMENTED** [review]
Not for this patch but I do wonder if there's value in protecting non-bf16 instruction backed builtins (e.g. loads, stores and shuffles) with the bf16 target guard.  I figure we'll either error on the use of the `svbfloat` type or the code generation should just work and thus there's no reason to artificially restrict user code.

File: clang/include/clang/Basic/arm_sve.td [hunk 1]
```diff
@@ -2264,6 +2278,18 @@ let TargetGuard = "sve2p1" in {
   defm SVPMOV_TO_VEC_LANE_D : PMOV_TO_VEC<"svpmov", "lUl", "aarch64_sve_pmov_to_vector_lane" ,[], ImmCheck1_7>;
 }
 
+let TargetGuard = "sve2p1|sme2" in {
```
[the remark this answers is not in the document]
Comment (sdesmalen-arm) [reply clang/include/clang/Basic/arm_sve.td]: Good spot!

File: clang/utils/TableGen/SveEmitter.cpp [hunk 2]
```diff
@@ -1781,7 +1781,12 @@ void SVEEmitter::createStreamingAttrs(raw_ostream &OS, ACLEKind Kind) {
   uint64_t VerifyRuntimeMode = getEnumValueForFlag("VerifyRuntimeMode");
   uint64_t IsStreamingCompatibleFlag =
       getEnumValueForFlag("IsStreamingCompatible");
+
   for (auto &Def : Defs) {
+    assert((((Def->getGuard().contains("sve") +
+              Def->getGuard().contains("sme")) <= 1) ||
+            Def->isFlagSet(VerifyRuntimeMode)) &&
```
[the remark this answers is not in the document]
Comment (sdesmalen-arm) [reply clang/utils/TableGen/SveEmitter.cpp]: Yes that makes sense. Done.

**Review by paulwalker-arm — COMMENTED** [review]
With the change of default it's very hard to check everything but we've already agreed  there'll need to be a full audit once all the inflight work has landed.  I did spot one thing though:

Should the integer svclamp and svrevd builtins be protected by "sve2p1|sme" rather than "sve2p1|sme2"?

File: clang/test/CodeGen/aarch64-sve-intrinsics/acle_sve_cntb.c [hunk 3]
```diff
@@ -17,7 +25,7 @@
 // CPP-CHECK-NEXT:    [[TMP1:%.*]] = shl nuw nsw i64 [[TMP0]], 4
 // CPP-CHECK-NEXT:    ret i64 [[TMP1]]
 //
-uint64_t test_svcntb()
+uint64_t test_svcntb(void) MODE_ATTR
```
[the remark this answers is not in the document]
Comment (sdesmalen-arm) [reply clang/test/CodeGen/aarch64-sve-intrinsics/acle_sve_cntb.c:28]: I don't think so.

The ACLE describes that the SME keyword attributes cannot be used with K&R unprototyped C functions, https://github.com/ARM-software/acle/blob/main/main/acle.md#sme-keyword-attributes

Clang gives a clear diagnostic describing the issue:  https://godbolt.org/z/7GabEWro1

**Review by paulwalker-arm — APPROVED** [review]

File: clang/test/CodeGen/aarch64-sve-intrinsics/acle_sve_cntb.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (aemerson) [reply clang/test/CodeGen/aarch64-sve-intrinsics/acle_sve_cntb.c:28]: Why is that restriction there? We often run into confusion when `int foo() __arm_streaming;` and the like is used.

**Pull request merged**